src/tournaments_watcher/notifier.rs

```diff
@@ -0,0 +1,140 @@
+use async_trait::async_trait;
+use serenity::model::prelude::ChannelId;
+use serenity::prelude::Context;
+
+use crate::data_watcher::DataUpdateNotifier;
+use crate::discord_utils::send_with_overflow;
+use crate::tournaments_watcher::ema::{
+    tournaments_diff, TournamentStatus, TournamentStatuses, Tournaments,
+};
+
+#[async_trait]
+pub trait TournamentsUpdateNotifier<R: Send + Sync> {
+    async fn notify(&self, ranking: &R);
+}
+
+pub struct TournamentsChannelMessageNotifier {
+    channel_id: ChannelId,
+    ctx: Context,
+    message: String,
+}
+
+impl TournamentsChannelMessageNotifier {
+    #[must_use]
+    pub fn new(channel_id: ChannelId, ctx: Context, message: String) -> Self {
+        Self {
+            channel_id,
+            ctx,
+            message,
+        }
+    }
+
+    #[must_use]
+    fn build_message(&self, tournament_statuses: &TournamentStatuses) -> String {
+        format!("{}{}", self.message, build_message(tournament_statuses))
+    }
+}
+
+#[must_use]
+fn build_message(tournaments: &TournamentStatuses) -> String {
+    let mut str = String::new();
+    for diff in &tournaments.0 {
+        str += &format!("* {}\n", diff_as_message(diff));
+    }
+
+    str
+}
+
+#[must_use]
+fn diff_as_message(diff: &TournamentStatus) -> String {
+    let mut str = String::new();
+
+    match diff {
+        TournamentStatus::New(entry) => {
+            str += "**NEW**: ";
+            str += &format!("_{}_", entry.name);
+            if !entry.url.is_empty() {
+                str += &format!(" ({})", entry.url);
+            }
+            str += "; ";
+            str += &format!("{}; ", entry.date);
+            str += &format!("{}; ", entry.place);
+            str += &format!("MERS: {}", entry.approval_status);
+            if !entry.results_status.is_empty() {
+                str += &format!("; {}", entry.results_status);
+            }
+        }
+        TournamentStatus::Changed(change) => {
+            str += &format!("**CHANGED**: _{}_; ", change.name);
+
+            if let Some(date) = &change.date {
+                str += &format!("date: {}; ", date);
+            }
+            if let Some(place) = &change.place {
+                str += &format!("place: {}; ", place);
+            }
+            if let Some(approval_status) = &change.approval_status {
+                str += &format!("MERS approval: {}; ", approval_status);
+            }
+            if let Some(results) = &change.results_status {
+                str += &format!("results: \"{}\"; ", results);
+            }
+
+            if str.ends_with("; ") {
+                str = str[0..str.len() - 2].to_owned();
+            }
```

Comment:
```suggestion
            if let Some(trimmed) = str.strip_suffix("; ") {
                str = trimmed.to_owned();
            }
```

Reply:
![image](https://github.com/riichi/chombot/assets/3128220/2c781bf3-a58c-4007-a8f9-9171a2deea3f)

Didn't think of it, noice.